libs/@local/hashql/core/src/intern/mod.rs

```diff
@@ -0,0 +1,132 @@
+mod map;
+mod set;
+use core::{
+    cmp::Ordering,
+    fmt::{self, Debug, Display},
+    hash::{Hash, Hasher},
+    ops::Deref,
+    ptr,
+};
+
+pub use self::{
+    map::{Decompose, InternMap, Provisioned},
+    set::InternSet,
+};
+
+mod private {
+    #[derive(Debug, Copy, Clone)]
+    pub struct Marker;
+}
+
+/// A reference to a value that is interned, and is known to be unique.
+///
+/// Note that it is possible to have a `T` and a `Interned<T>` that are (or
+/// refer to) equal but different values. But if you have two different
+/// `Interned<T>`s, they both refer to the same value, at a single location in
+/// memory. This means that equality and hashing can be done on the value's
+/// address rather than the value's contents, which can improve performance.
+///
+/// The `Marker` field means you can pattern match with `Interned(v, _)`
+/// but you can only construct a `Interned` with `new_unchecked`, and not
+/// directly.
+pub struct Interned<'heap, T: ?Sized>(pub &'heap T, pub private::Marker);
+
+impl<'heap, T: ?Sized> Interned<'heap, T> {
+    /// Create a new `Interned` value.
+    ///
+    /// The value referred to *must* be interned
+    /// and thus be unique, and it *must* remain unique in the future. This
+    /// function has `_unchecked` in the name but is not `unsafe`, because if
+    /// the uniqueness condition is violated condition it will cause incorrect
+    /// behaviour but will not affect memory safety.
+    #[inline]
+    pub const fn new_unchecked(value: &'heap T) -> Self {
+        Interned(value, private::Marker)
+    }
+}
+
+impl<T: ?Sized + Debug> Debug for Interned<'_, T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Debug::fmt(self.0, fmt)
+    }
+}
+
+impl<T: ?Sized + Display> Display for Interned<'_, T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Display::fmt(self.0, fmt)
+    }
+}
+
+impl<T: ?Sized> Clone for Interned<'_, T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T: ?Sized> Copy for Interned<'_, T> {}
+
+impl<T: ?Sized> Deref for Interned<'_, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl<T: ?Sized> AsRef<T> for Interned<'_, T> {
+    fn as_ref(&self) -> &T {
+        self.0
+    }
+}
+
+impl<T: ?Sized> PartialEq for Interned<'_, T> {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        // Pointer equality implies equality, due to the uniqueness constraint.
+        ptr::eq(self.0, other.0)
+    }
+}
+
+impl<T: ?Sized> Eq for Interned<'_, T> {}
+
+impl<T: ?Sized + PartialOrd> PartialOrd for Interned<'_, T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        // Pointer equality implies equality, due to the uniqueness constraint,
```

Comment:
What do you mean with "Uniqueness constraint"? We have a reference, if the reference is the same, it's equal, yes, but that should be implemented by Rust itself already.

Reply:
The uniqueness constraint is elaborated above, the value itself is guaranteed to be unique, due to the construction rules (see e.g. `new_unchecked`), this means that instead of doing a comparison of the inner value we can simply compare pointers, therefore being a lot faster than comparing the contents.

https://github.com/hashintel/hash/blob/f037b1db96ad3a8243d860bd26eda7fb61afce79/libs/%40local/hashql/core/src/intern/mod.rs#L35-L41

Reply:
But this only holds types held by the same interner, right?
Also, as Rust forbids aliasing, I don't see why we need to special handle this case at all. If it compares two references and the pointers are equal, it's guaranteed that both values are the same. This is something, rustc can optimize itself.

Reply:
Okay, in fact `Rust` does not